pages/building-packages-for-runtimes.md

```diff
@@ -0,0 +1,239 @@
+    "default": "./source/vendor/supports-color/browser.js"
+  }
+}
+```
+
+<img style="width: auto; height: 400px" src="./chalk.gif" alt="Gif to show generation of import-map for chalk. (package)" />
+
+> `https://ga.jspm.io/npm:chalk@5.2.0/source/vendor/supports-color/index.js` > `https://ga.jspm.io/npm:chalk@5.2.0/source/vendor/supports-color/browser.js`
+
+## Exports
+
+In addition to the `imports` field, there is also an `exports` field in the `package.json` file that allows package authors to specify how their package should be imported in different environments. The `exports` field allows authors to specify different entry points for different runtimes(browser, node, default), as well as different formats (such as CommonJS or ES modules) for the same entry point. This allows for more efficient loading and usage of packages in different environments.
+
+Don’t get confused between `imports` and `exports`. Imports are for using runtime aware external packages into your module. Whereas `exports` are used to expose your module in different environments making it work across runtimes and formats.
+
+A example of exports filed can be found in
+
+[https://github.com/preactjs/preact/blob/master/package.json](https://github.com/preactjs/preact/blob/master/package.json)
+
+```json
+"type": "module",
+"exports": {
+  ".": {
+    "types": "./src/index.d.ts",
+    "browser": "./dist/preact.module.js",
+    "umd": "./dist/preact.umd.js",
+    "import": "./dist/preact.mjs",
+    "require": "./dist/preact.js"
+  }
+}
+```
+
+**Open specifications**
+
+[WinterCG](https://wintercg.org/) the collaboration platform for different javascript runtimes is in the process of standardising these runtime keys. Here is the [specification](https://runtime-keys.proposal.wintercg.org/) for more details. The proposed specification tries to bring in more identifiers for the different runtimes. This helps in optimising the packages by taking advantages of the built-ins that are available in the target runtime. And adding polyfills for those which are only needed.
+
+## ENV
+
+While bundling for projects, it’s a common practice that we use `process.env.NODE_ENV` to have different builds for `production` and `development`. Well, this is common practice in building applications, but when packaging for libraries, this becomes a problem. As `process` is note a browser built-in. It is of the best interest for the authors of these libraries to serve different versions in different build targets to give better error messages. Better build sizes etc. Trying load these in browsers directly from CDN’s crashes the module.
+
+One example is, by trying load `react-router@5.2.0` into the browser as `esm` module. Fails to load it, as the `entry` uses `process.env.NODE_ENV`
+
+[https://unpkg.com/browse/react-router@5.2.0/index.js](https://unpkg.com/browse/react-router@5.2.0/index.js)
+
+```jsx
+// react-router/index.js
+if (process.env.NODE_ENV === "production") {
```

Comment:
This is only handled in CommonJS and is generally discouraged! It is much better to use conditional exports and imports for branching rather than code anaylsis.